include/larch/impl/dagbin_fileio_impl.hpp

```diff
@@ -0,0 +1,315 @@
+
+MADAGStorage<> DagbinFileIO::ReadDAG(std::string_view path) {
+  std::ifstream infile(std::string{path}, std::ios::binary);
+
+  MADAGStorage<> dag_storage = MADAGStorage<>::EmptyDefault();
+  auto dag = dag_storage.View();
+  Header header;
+
+  while (!infile.eof()) {
+    auto append_id = ReadData<char>(infile);
+    auto nxt_offset = ReadData<std::streampos>(infile);
+    if (infile.eof()) {
+      break;
+    }
+
+    std::string test_str;
+    switch (append_id) {
+      case 'H':
+        header = ReadHeader(infile, dag);
+        break;
+      case 'R':
+        ReadReferenceSequence(infile, dag);
+        break;
+      case 'N':
+        ReadNodes(infile, dag);
+        break;
+      case 'E':
+        ReadEdges(infile, dag);
+        break;
+      case 'X':
+        test_str = ReadString(infile);
+        std::cout << ">>> READ test: '" << test_str << "'" << std::endl;
+        break;
+      default:
+        std::cerr << "ERROR: Invalid append_id '" << append_id << "'" << std::endl;
+        std::exit(EXIT_FAILURE);
+    }
+    Assert(nxt_offset == infile.tellg());
+  }
+  infile.close();
+
+  dag.BuildConnections();
+  dag.AssertUA();
+  for (auto node : dag.GetNodes()) {
```

Comment:
There still needs to be a dag.RecomputeCompactGenomes() call before using the CGs to populate SampleIds. Nodes are written without cgs, and are recovered from the edge mutations. Therefore this loop that creates a SampleId based on CGs will not do what you expect.